crates/client/rpc/src/lib.rs

```diff
@@ -681,13 +708,128 @@ where
     }
 
     /// Returns all events matching the given filter
-    async fn get_events(
-        &self,
-        filter: EventFilter,
-        continuation_token: Option<String>,
-        chunk_size: u64,
-    ) -> RpcResult<EventsPage> {
-        todo!("Not implemented")
+    async fn get_events(&self, filter: EventFilterWithPage) -> RpcResult<EventsPage> {
+        let mut continuation_token = match filter.result_page_request.continuation_token {
+            Some(token) => token.parse::<usize>().map_err(|e| {
+                error!("Failed to parse continuation token: {:?}", e);
+                StarknetRpcApiError::InvalidContinuationToken
+            })?,
+            None => 0usize,
+        };
+        let filter_address = filter.event_filter.address.map(Felt252Wrapper::from);
+        let keys = filter.event_filter.keys.unwrap_or_default();
+        let chunk_size = filter.result_page_request.chunk_size;
+
+        if keys.len() > MAX_EVENTS_KEYS {
+            return Err(StarknetRpcApiError::TooManyKeysInFilter.into());
+        }
+        if chunk_size > MAX_EVENTS_CHUNK_SIZE {
+            return Err(StarknetRpcApiError::PageSizeTooBig.into());
+        }
+
+        // Get the substrate block numbers for the requested range
+        let latest_block =
+            self.substrate_block_number_from_starknet_block(BlockId::Tag(BlockTag::Latest)).map_err(|e| {
+                error!("'{e}'");
+                StarknetRpcApiError::BlockNotFound
+            })?;
+        let mut current_block = self
```

Comment:
To confirm:
`from_block` default implementation is to start at genesis if not provided?

Reply:
I couldn't find a spec which says this explicitly but this is the behaviour followed by [Papyrus](https://github.com/starkware-libs/papyrus/blob/main/crates/papyrus_gateway/src/api/v0_3_0/v0_3_0_impl.rs#L347).